### PR TITLE
[MUX-1375] - Changes to GSPlayer 

### PR DESCRIPTION
**Pull request by gurpreet-ios**
1. Added functionality to check if the video is being played local
2. Refactored code according to previous implementation.

### CODE REVIEW

File: GSPlayer/Classes/Cache/VideoCacheHandler.swift [hunk 1]
```diff
@@ -127,11 +127,13 @@ public class VideoCacheHandler {
     }
     
     func cache(data: Data, for range: NSRange) {
-        objc_sync_enter(writeFileHandle)
-        writeFileHandle.seek(toFileOffset: UInt64(range.location))
-        writeFileHandle.write(data)
-        configuration.add(fragment: range)
-        objc_sync_exit(writeFileHandle)
+        DispatchQueue.global(qos: .default).async {
```
Comment (pratik-gits) [inline GSPlayer/Classes/Cache/VideoCacheHandler.swift:130]: add comment why is this needed.

File: GSPlayer/Classes/Cache/VideoCacheHandler.swift [hunk 2]
```diff
@@ -143,11 +145,13 @@ public class VideoCacheHandler {
     }
     
     func set(info: VideoInfo) {
-        objc_sync_enter(writeFileHandle)
-        configuration.info = info
-        writeFileHandle.truncateFile(atOffset: UInt64(info.contentLength))
-        writeFileHandle.synchronizeFile()
-        objc_sync_exit(writeFileHandle)
+        self.configuration.info = info
```
Comment (pratik-gits) [inline GSPlayer/Classes/Cache/VideoCacheHandler.swift:148]: add comment why is this needed.

File: GSPlayer/Classes/Download/VideoDownloader.swift [hunk 3]
```diff
@@ -33,6 +33,12 @@ public class VideoDownloader {
     }
     
     public func downloadToEnd(from offset: Int) {
+        if offset == 0 {
```
Comment (pratik-gits) [inline GSPlayer/Classes/Download/VideoDownloader.swift:36]: add comment why is this needed.

File: GSPlayer/Classes/Download/VideoDownloaderHandler.swift [hunk 4]
```diff
@@ -171,30 +169,28 @@ private extension VideoDownloaderHandler {
...
-        let configuration = cacheHandler.configuration
-        NotificationCenter.default.post(
-            name: .VideoDownloadProgressDidChanged,
-            object: nil,
-            userInfo: ["configuration": configuration]
-        )
+}
+
+
+private extension UIDevice {
```
Comment (pratik-gits) [inline GSPlayer/Classes/Download/VideoDownloaderHandler.swift]: move this to a separate file

File: GSPlayer/Classes/Extension/AVPlayerItem+Extensions.swift [hunk 5]
```diff
@@ -42,6 +42,14 @@ public extension AVPlayerItem {
         return urlString.replacingOccurrences(of: AVPlayerItem.loaderPrefix, with: "").url
     }
     
+    var hasVideoCached: Bool {
```
Comment (pratik-gits) [inline GSPlayer/Classes/Extension/AVPlayerItem+Extensions.swift]:
```suggestion
    var isVideoAlreadyCached: Bool {
```

File: GSPlayer/Classes/View/VideoPlayerView.swift [hunk 6]
```diff
@@ -118,6 +122,11 @@ open class VideoPlayerView: UIView {
         return isLoaded ? currentDuration + totalDuration * Double(replayCount) : 0
     }
     
+    /// Returns true if the video is being played from local
+    public var isVideoPlayingLocally:Bool {
```
Comment (pratik-gits) [inline GSPlayer/Classes/View/VideoPlayerView.swift]: change name to something else

File: GSPlayer/Classes/View/VideoPlayerView.swift [hunk 7]
```diff
@@ -346,7 +373,7 @@ private extension VideoPlayerView {
                 self.state = .paused(playProgress: self.playProgress, bufferProgress: self.bufferProgress)
             }
             
-            if self.bufferProgress >= 0.99 || (self.currentBufferDuration - self.currentDuration) > 3 {
+            if self.bufferProgress >= 0.99 || (self.currentBufferDuration - self.currentDuration) > 5.5 {
```
Comment (pratik-gits) [inline GSPlayer/Classes/View/VideoPlayerView.swift:377]: check what this 5.5 actually is

**Review by pratik-gits — COMMENTED** [review]
Clarification required

File: GSPlayer/Classes/View/VideoPlayerView.swift (hunk 7 above)
Comment (gurpreet-ios) [reply GSPlayer/Classes/View/VideoPlayerView.swift:377]: added the description as per logic

**Review by pratik-gits — APPROVED** [review]

**Review by pratik-gits — APPROVED** [review]

**Review by omkar-sharechat — APPROVED** [review]
